pysipp/agent.py

```diff
@@ -444,13 +445,27 @@ def from_agents(self, agents=None, autolocalsocks=True, **scenkwargs):
 
     async def arun(
         self,
-        timeout=180,
+        timeout=DEFAULT_RUNNER_TIMEOUT,
         runner=None,
+        block=True,
     ):
-        agents = self.prepare()
-        runner = runner or launch.TrioRunner()
+        self._prepared_agents = agents = self.prepare()
+        self._runner = runner = runner or launch.TrioRunner()
 
-        return await launch.run_all_agents(runner, agents, timeout=timeout)
+        return await launch.run_all_agents(runner, agents, timeout=timeout, block=block)
+
+    def finalize(self, *, timeout=DEFAULT_RUNNER_TIMEOUT):
+        assert (
+            self._prepared_agents and self._runner
+        ), "Must run scenario before finalizing."
+        return trio.run(
```

Comment:
Ahh right this is triggered from sync code yes?

Sorry trying to get my head back in this change set!

Reply:
yes, it's not the most elegant solution, do you want me to try something that doesn't require abusing objects attributes and be more functional? (it may change the API a bit though)

Reply:
If you have something functional that you think is cleaner then yeah for sure.

I have no real problems with this I just forgot this is how we have to do it.

Reply:
maybe we can just return a handle to finalize from `Scenario.run()`, atm we return the runner which is exposing internals.

Reply:
Yeah this is a good idea - feels like a continuation/future then.